File: utils/parser.py
import re
import logging

logger = logging.getLogger(__name__)

def parse_vss110_text(text: str) -> list:
    """
    Parses a Visa Settlement text report and extracts VSS-110 report data.
    
    Args:
        text (str): The raw text content of the report file.
        
    Returns:
        list: A list of dictionaries, one for each parsed VSS-110 report section.
    """
    # 1. Extract independent VSS-110 report sections
    # Using re.DOTALL to let the dot . match newline characters.
    section_pattern = re.compile(
        r'(REPORT ID:\s+VSS-110.*?\*\*\*\s+END OF VSS-110 REPORT\s+\*\*\*)', 
        re.DOTALL
    )
    
    sections = section_pattern.findall(text)
    logger.info(f"Found {len(sections)} potential VSS-110 report sections in file.")
    
    parsed_reports = []
    
    for idx, sec in enumerate(sections, 1):
        try:
            # 2. Extract standard header information
            # Reporting For
            reporting_for_match = re.search(r'REPORTING FOR:\s*([^\n\r]+)', sec)
            reporting_for = reporting_for_match.group(1).strip() if reporting_for_match else "UNKNOWN"
            
            # Report Date
            report_date_match = re.search(r'REPORT DATE:\s*([^\s\n\r]+)', sec)
            report_date = report_date_match.group(1).strip() if report_date_match else "UNKNOWN"
            
            # Settlement Currency
            settlement_currency_match = re.search(r'SETTLEMENT CURRENCY:\s*([A-Za-z0-9]+)', sec)
            settlement_currency = settlement_currency_match.group(1).strip() if settlement_currency_match else "UNKNOWN"
            
            # 3. Extract Reimbursement Fees Section data
            # First focus on the section under "REIMBURSEMENT FEES" up to "TOTAL OTHER" / next header
            reimb_section_match = re.search(r'REIMBURSEMENT FEES.*?(?:TOTAL ISSUER\s+([\d,]+\.\d+)\s+([\d,]+\.\d+)\s+([\d,]+\.\d+(?:CR|DB)?))', sec, re.DOTALL)
            
            reimb_credit = "0.00"
            reimb_debit = "0.00"
            reimb_total = "0.00"
            
            if reimb_section_match:
                reimb_credit = reimb_section_match.group(1)
                reimb_debit = reimb_section_match.group(2)
                reimb_total = reimb_section_match.group(3)
            else:
                # Fallback search if the total issuer is further down
                reimb_sec_block = re.search(r'REIMBURSEMENT FEES.*?(?=VISA CHARGES|TOTAL OTHER|$)', sec, re.DOTALL)
                if reimb_sec_block:
                    reimb_match = re.search(r'TOTAL ISSUER\s+([\d,]+\.\d+)\s+([\d,]+\.\d+)\s+([\d,]+\.\d+(?:CR|DB)?)', reimb_sec_block.group(0))
                    if reimb_match:
                        reimb_credit = reimb_match.group(1)
                        reimb_debit = reimb_match.group(2)
                        reimb_total = reimb_match.group(3)

            # 4. Extract Visa Charges Section data
            # Locate first the block starting with "VISA CHARGES" and extract "TOTAL ISSUER"
            # It must ignore "TOTAL OTHER"
            visa_section_match = re.search(r'VISA CHARGES.*?(?:TOTAL ISSUER\s+([\d,]+\.\d+)\s+([\d,]+\.\d+)\s+([\d,]+\.\d+(?:CR|DB)?))', sec, re.DOTALL)
            
            visa_credit = "0.00"
            visa_debit = "0.00"
            visa_total = "0.00"
            
            if visa_section_match:
                visa_credit = visa_section_match.group(1)
                visa_debit = visa_section_match.group(2)
                visa_total = visa_section_match.group(3)
            else:
                visa_sec_block = re.search(r'VISA CHARGES.*?(?=\*\*\* END|$)', sec, re.DOTALL)
                if visa_sec_block:
                    visa_match = re.search(r'TOTAL ISSUER\s+([\d,]+\.\d+)\s+([\d,]+\.\d+)\s+([\d,]+\.\d+(?:CR|DB)?)', visa_sec_block.group(0))
                    if visa_match:
                        visa_credit = visa_match.group(1)
                        visa_debit = visa_match.group(2)
                        visa_total = visa_match.group(3)

            # Assemble parsed record
            record = {
                "REPORT_ID": "VSS-110",
                "REPORTING_FOR": reporting_for,
                "REPORT_DATE": report_date,
                "SETTLEMENT_CURRENCY": settlement_currency,
                "REIMBURSEMENT_FEES_CREDIT": reimb_credit,
                "REIMBURSEMENT_FEES_DEBIT": reimb_debit,
                "REIMBURSEMENT_FEES_TOTAL": reimb_total,
                "VISA_CHARGES_CREDIT": visa_credit,
                "VISA_CHARGES_DEBIT": visa_debit,
                "VISA_CHARGES_TOTAL": visa_total
            }
            parsed_reports.append(record)
            
        except Exception as e:
            logger.error(f"Error parsing section index {idx}: {str(e)}")
            continue
            
    return parsed_reports

File: src/utils/parser.ts
export interface VSS110Record {
  REPORT_ID: string;
  REPORTING_FOR: string;
  REPORT_DATE: string;
  SETTLEMENT_CURRENCY: string;
  REIMBURSEMENT_FEES_CREDIT: string;
  REIMBURSEMENT_FEES_DEBIT: string;
  REIMBURSEMENT_FEES_TOTAL: string;
  VISA_CHARGES_CREDIT: string;
  VISA_CHARGES_DEBIT: string;
  VISA_CHARGES_TOTAL: string;
}

/**
 * Parses a Visa Settlement text report and extracts VSS-110 report data.
 * Matches the python-based regex rules and handles edge cases beautifully.
 */
export function parseVSS110Text(text: string): VSS110Record[] {
  // 1. Isolate VSS-110 report sections
  // Using 'g' for global matching and 's' for dotAll (so dot matches newlines)
  const sectionRegex = /REPORT ID:\s+VSS-110.*?\*\*\*\s+END OF VSS-110 REPORT\s+\*\*\*/gs;
  const sections = text.match(sectionRegex) || [];
  
  const parsedRecords: VSS110Record[] = [];

  sections.forEach((sec, idx) => {
    try {
      // Extract reporting identity details
      const reportingForMatch = sec.match(/REPORTING FOR:\s*([^\n\r]+)/);
      const reportingFor = reportingForMatch ? reportingForMatch[1].trim() : "UNKNOWN";

      const reportDateMatch = sec.match(/REPORT DATE:\s*([^\s\n\r]+)/);
      const reportDate = reportDateMatch ? reportDateMatch[1].trim() : "UNKNOWN";

      const currencyMatch = sec.match(/SETTLEMENT CURRENCY:\s*([A-Za-z0-9]+)/);
      const settlementCurrency = currencyMatch ? currencyMatch[1].trim() : "UNKNOWN";

      // Extract REIMBURSEMENT FEES
      // Matches the python rules: REIMBURSEMENT FEES followed by TOTAL ISSUER and credit/debit/total
      const reimbRegex = /REIMBURSEMENT FEES.*?TOTAL ISSUER\s+([\d,]+\.\d+)\s+([\d,]+\.\d+)\s+([\d,]+\.\d+(?:CR|DB)?)/s;
      const reimbMatch = sec.match(reimbRegex);

      let reimbCredit = "0.00";
      let reimbDebit = "0.00";
      let reimbTotal = "0.00";

      if (reimbMatch) {
        reimbCredit = reimbMatch[1];
        reimbDebit = reimbMatch[2];
        reimbTotal = reimbMatch[3];
      } else {
        // Fallback: search specifically inside the REIMBURSEMENT FEES block
        const reimbBlockMatch = sec.match(/REIMBURSEMENT FEES.*?(?=VISA CHARGES|TOTAL OTHER|$)/s);
        if (reimbBlockMatch) {
          const blockText = reimbBlockMatch[0];
          const fallbackMatch = blockText.match(/TOTAL ISSUER\s+([\d,]+\.\d+)\s+([\d,]+\.\d+)\s+([\d,]+\.\d+(?:CR|DB)?)/);
          if (fallbackMatch) {
            reimbCredit = fallbackMatch[1];
            reimbDebit = fallbackMatch[2];
            reimbTotal = fallbackMatch[3];
          }
        }
      }

      // Extract VISA CHARGES (specifically isolating TOTAL ISSUER block in it)
      const visaRegex = /VISA CHARGES.*?TOTAL ISSUER\s+([\d,]+\.\d+)\s+([\d,]+\.\d+)\s+([\d,]+\.\d+(?:CR|DB)?)/s;
      const visaMatch = sec.match(visaRegex);

      let visaCredit = "0.00";
      let visaDebit = "0.00";
      let visaTotal = "0.00";

      if (visaMatch) {
        visaCredit = visaMatch[1];
        visaDebit = visaMatch[2];
        visaTotal = visaMatch[3];
      } else {
        // Fallback: search specifically inside the VISA CHARGES block
        const visaBlockMatch = sec.match(/VISA CHARGES.*?(?=\*\*\* END|$)/s);
        if (visaBlockMatch) {
          const blockText = visaBlockMatch[0];
          const fallbackMatch = blockText.match(/TOTAL ISSUER\s+([\d,]+\.\d+)\s+([\d,]+\.\d+)\s+([\d,]+\.\d+(?:CR|DB)?)/);
          if (fallbackMatch) {
            visaCredit = fallbackMatch[1];
            visaDebit = fallbackMatch[2];
            visaTotal = fallbackMatch[3];
          }
        }
      }

      parsedRecords.push({
        REPORT_ID: "VSS-110",
        REPORTING_FOR: reportingFor,
        REPORT_DATE: reportDate,
        SETTLEMENT_CURRENCY: settlementCurrency,
        REIMBURSEMENT_FEES_CREDIT: reimbCredit,
        REIMBURSEMENT_FEES_DEBIT: reimbDebit,
        REIMBURSEMENT_FEES_TOTAL: reimbTotal,
        VISA_CHARGES_CREDIT: visaCredit,
        VISA_CHARGES_DEBIT: visaDebit,
        VISA_CHARGES_TOTAL: visaTotal
      });
    } catch (e) {
      console.error(`Error parsing VSS-110 section ${idx}:`, e);
    }
  });

  return parsedRecords;
}
